### PR TITLE
fix(#1013): adding new diagnostic for missing expression on ${} (missing expression)

### PR DESCRIPTION
**Pull request by HicaroD**

### CODE REVIEW

**Comment by HicaroD** [comment]
I have read and hereby agree to quick-lint-js' Contributor License Agreement (CLA-v1.md).

File: docs/errors/E0711.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+Everytime you use `${}` within a template literal, which is a string defined
+with backticks, such as this one below, you should specify an expression
+inside its curly braces.
+
+```javascript
+const myString = `Here is a string defined with backticks`;
+```
+
+See an example:
+
```
Comment (strager) [inline docs/errors/E0711.md:12]: Must fix: The first code example in the file should demonstrate the error. Only your second code example demonstrates the error.
```suggestion
```

File: docs/errors/E0711.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+```javascript
+console.log(`My name is ${}`); // error: missing expression in placeholder within template literal
```
Comment (strager) [inline docs/errors/E0711.md:14]: Must fix: Do not repeat the error message. The website generated from this file will highlight the error automatically.
```suggestion
console.log(`My name is ${}`);
```

File: docs/errors/E0711.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+```
+
+Here is how you can solve it:
+
+```javascript
+const yourName = "Ryan";
+console.log(`My name is ${yourName}`); // Correct
```
Comment (strager) [inline docs/errors/E0711.md:21]: Nit: There is no need to say that the working example is correct.
```suggestion
console.log(`My name is ${yourName}`);
```

File: docs/errors/E0711.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+const myString = `Here is a string defined with backticks`;
+```
+
+See an example:
+
+```javascript
+console.log(`My name is ${}`); // error: missing expression in placeholder within template literal
+```
+
+Here is how you can solve it:
```
Comment (strager) [inline docs/errors/E0711.md:17]: You should explain *and* show, not just show.
```suggestion
To fix this error, write an expression between `${` and `}`:
```

File: docs/errors/E0711.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+```javascript
+const yourName = "Ryan";
+console.log(`My name is ${yourName}`); // Correct
+```
```
Comment (strager) [inline docs/errors/E0711.md:22]: It is possible that the programmer intended to write `\${}` instead of `${}`. I think it's useful to suggest escaping the `$` or using `""`/`''` instead of backticks.

File: src/quick-lint-js/diag/diagnostic-types-2.h [hunk 6]
```diff
@@ -2937,6 +2937,13 @@ struct Diag_Xor_Used_As_Exponentiation {
                   ARG(xor_operator))]]  //
   Source_Code_Span xor_operator;
 };
+
+struct Diag_Expected_Expression_In_Template_Literal {
+  [[qljs::diag("E0711", Diagnostic_Severity::error)]]  //
+  [[qljs::message("missing expression in placeholder within template literal",
+                  ARG(template_literal))]]  //
+  Source_Code_Span template_literal;
```
Comment (strager) [inline src/quick-lint-js/diag/diagnostic-types-2.h:2945]: This should be named `placeholder`, not `template_literal`.

File: test/test-lex.cpp [hunk 7]
```diff
@@ -1017,6 +1017,11 @@ world`)"_sv,
       u8"`unterminated\\"_sv,  //
       u8"^^^^^^^^^^^^^^^ Diag_Unclosed_Template"_diag,
       {Token_Type::complete_template});
+
+  this->check_tokens_with_errors(
+      u8"`${}`"_sv,  //
+      u8"  ^^ Diag_Expected_Expression_In_Template_Literal"_diag,
+      {Token_Type::complete_template});
```
Comment (strager) [inline test/test-lex.cpp:1024]: Please add a test for `${ }` (whitespace inside). I think this test would fail.

File: test/test-lex.cpp [hunk 8]
```diff
@@ -1017,6 +1017,11 @@ world`)"_sv,
       u8"`unterminated\\"_sv,  //
       u8"^^^^^^^^^^^^^^^ Diag_Unclosed_Template"_diag,
       {Token_Type::complete_template});
+
+  this->check_tokens_with_errors(
+      u8"`${}`"_sv,  //
+      u8"  ^^ Diag_Expected_Expression_In_Template_Literal"_diag,
```
Comment (strager) [inline test/test-lex.cpp:1023]: Nit: I haven't thought about this much, but perhaps it is better to put the span between the `{}`:
```suggestion
      u8"   ` Diag_Expected_Expression_In_Template_Literal"_diag,
```

File: test/test-lex.cpp (hunk 7 above)
Comment (strager) [inline test/test-lex.cpp:1024]: Please add a test for \`a${}b${c}\`. I want to ensure that the template is parsed correctly after error recovery.

**Review by strager — CHANGES_REQUESTED** [review]

File: docs/errors/E0711.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+If your goals were to have it literally in your string, consider escaping it
+using a backslash before the placeholder:
+
+```javascript
+console.log(`\${}`);
+```
+
+Or just use double-quoted string (`""`):
```
Comment (strager) [inline docs/errors/E0711.md:25]: Nit: Avoid using the word "just".
```suggestion
Or use double-quoted string (`""`):
```

File: docs/errors/E0711.md [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+To fix this error, write an expression between `${` and `}`:
+
+```javascript
+const yourName = "Ryan";
+console.log(`My name is ${yourName}`);
+```
+
+If your goals were to have it literally in your string, consider escaping it
+using a backslash before the placeholder:
```
Comment (strager) [inline docs/errors/E0711.md]: Nit: What is "it"?

File: src/quick-lint-js/fe/parse-expression.cpp [hunk 11]
```diff
@@ -3964,6 +3963,12 @@ Expression* Parser::parse_untagged_template(Parse_Visitor_Base& v) {
     this->peek().report_errors_for_escape_sequences_in_template(
         this->diag_reporter_);
     this->skip();
+
+    if (this->peek().type == Token_Type::right_curly) {
```
Comment (strager) [inline src/quick-lint-js/fe/parse-expression.cpp:3967]: Minor: If this is true, we shouldn't call `parse_expression` (line 3972).

File: test/test-parse-expression.cpp [hunk 12]
```diff
@@ -1711,6 +1711,43 @@ TEST_F(Test_Parse_Expression, optional_tagged_template_literal) {
...
+TEST_F(Test_Parse_Expression, Diag_Expected_Expression_In_Template_Literal) {
+  {
+    Test_Parser p(u8"`${}`"_sv, capture_diags);
+    p.parse_and_visit_expression();
+    assert_diagnostics(
+        p.code, p.errors,
+        {
+            u8"Diag_Expected_Expression_In_Template_Literal"_diag,
+        });
+  }
```
Comment (strager) [inline test/test-parse-expression.cpp:1723]: Use `test_parse_and_visit_expression` instead. It'll be much shorter.

File: test/test-parse-expression.cpp [hunk 13]
```diff
@@ -1711,6 +1711,43 @@ TEST_F(Test_Parse_Expression, optional_tagged_template_literal) {
...
 }
 
+TEST_F(Test_Parse_Expression, Diag_Expected_Expression_In_Template_Literal) {
+  {
+    Test_Parser p(u8"`${}`"_sv, capture_diags);
+    p.parse_and_visit_expression();
+    assert_diagnostics(
+        p.code, p.errors,
+        {
+            u8"Diag_Expected_Expression_In_Template_Literal"_diag,
```
Comment (strager) [inline test/test-parse-expression.cpp:1721]: Must fix: Include location information in the assertion (leading `^^^` or \`). I think your implementation currently puts the diagnostic in a weird spot.

**Review by strager — CHANGES_REQUESTED** [review]

**Review by strager — APPROVED** [review]

**Comment by strager** [comment]
Landed as commit 67c6c67e51ebed914e7a5c5e9e40b9b067f8ff3e. Thanks for the fix!
